File: frontend/packages/types/src/api-error.ts
export interface FieldError {
  field: string;
  message: string;
}

/** Uniform error body returned by gateway and backend services. */
export interface ApiError {
  code: string;
  message: string;
  traceId: string;
  timestamp: string;
  fieldErrors?: FieldError[];
}

export type ApiErrorCode =
  | 'ACCOUNT_NOT_ACTIVE'
  | 'FORBIDDEN'
  | 'USER_STATUS_UNAVAILABLE'
  | 'MISSING_TOKEN'
  | 'INVALID_TOKEN'
  | string;

File: platform/parkio-platform/src/main/java/com/parkio/platform/api/ApiError.java
package com.parkio.platform.api;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.parkio.platform.http.PlatformHeaders;
import java.time.Instant;
import java.util.List;
import org.slf4j.MDC;

/**
 * Service-agnostic API error envelope.
 *
 * <p>This type is infrastructure-only. Services still own exception mapping and
 * status-code decisions.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public record ApiError(
        String code,
        String message,
        String traceId,
        List<FieldError> fieldErrors,
        Instant timestamp) {

    public record FieldError(String field, String message) {
    }

    public static ApiError of(String code, String message, Instant timestamp) {
        return new ApiError(code, message, currentTraceId(), null, timestamp);
    }

    public static ApiError of(String code, String message, List<FieldError> fieldErrors, Instant timestamp) {
        return new ApiError(code, message, currentTraceId(), fieldErrors, timestamp);
    }

    private static String currentTraceId() {
        String traceId = MDC.get(PlatformHeaders.MDC_CORRELATION_ID);
        return traceId == null || traceId.isBlank() ? "unknown" : traceId;
    }
}
